inc/Engine/CriticalPath/CriticalCSSSubscriber.php

```diff
@@ -687,39 +594,82 @@ public function async_css( $buffer ) {
 		$css_pattern = apply_filters(
 			'rocket_async_css_regex_pattern',
-			'/(?=<link[^>]*\s(rel\s*=\s*[\'"]stylesheet["\']))<link[^>]*\shref\s*=\s*[\'"]([^\'"]+)[\'"](.*)>/iU'
+			'/(?=<link[^>]*\s(rel\s*=\s*[\'"]stylesheet["\']))<link(?=[^<>]*?(?:(?<type>type\s*=\s*[\'"][^<>"\']*[\'"])|>))(?=[^<>]*?(?:(?<onload>onload\s*=\s*(.*))|>))(?=[^<>]*?(?:(?<media>media\s*=\s*[\'"][^<>"\']*[\'"])|>))(?=[^<>]*?(?:href\s*=\s*[\'"](?<url>.*?)[\'"]|>))(?=[^<>]*?(?:(?<rel>rel\s*=\s*[\'"]>[^<>"\']*)[\'"]|>))(?:.*?<\/\1>|[^<>]*>)/ix'
 		);
 
 		// Get all css files with this regex.
 		preg_match_all( $css_pattern, $buffer, $tags_match );
+
 		if ( ! isset( $tags_match[0] ) ) {
 			return $buffer;
 		}
 
-		$noscripts = '';
+		return $this->async_css_tag_update( $tags_match, $buffer );
+	}
 
+	/**
+	 * Updates the HTML code by defering all CSS matches files.
+	 *
+	 * @param  array  $tags_match All matched css files.
+	 * @param  string $buffer     HTML code.
+	 * @return string             Modified HTML code with defer CSS.
+	 */
+	private function async_css_tag_update( $tags_match, $buffer ) {
+		$excluded_css = array_flip( $this->critical_css->get_exclude_async_css() );
+		$noscripts    = '';
 		foreach ( $tags_match[0] as $i => $tag ) {
 			// Strip query args.
-			$path = rocket_extract_url_component( $tags_match[2][ $i ], PHP_URL_PATH );
+			$path = rocket_extract_url_component( $tags_match['url'][ $i ], PHP_URL_PATH );
 
 			// Check if this file should be deferred.
 			if ( isset( $excluded_css[ $path ] ) ) {
 				continue;
 			}
 
-			$preload = str_replace( 'stylesheet', 'preload', $tags_match[1][ $i ] );
-			$onload  = preg_replace( '~' . preg_quote( $tags_match[3][ $i ], '~' ) . '~iU', ' as="style" onload=""' . $tags_match[3][ $i ] . '>', $tags_match[3][ $i ] );
-			$tag     = str_replace( $tags_match[3][ $i ] . '>', $onload, $tag );
-			$tag     = str_replace( $tags_match[1][ $i ], $preload, $tag );
-			$tag     = str_replace( 'onload=""', 'onload="this.onload=null;this.rel=\'stylesheet\'"', $tag );
-			$buffer  = str_replace( $tags_match[0][ $i ], $tag, $buffer );
+			$original_media = 'all';
+			$tag            = str_replace( $tags_match['type'][ $i ], ' as="style" ' . $tags_match['type'][ $i ] . ( empty( $tags_match['media'][ $i ] ) ? ' media="print"' : '' ) . ( empty( $tags_match['onload'][ $i ] ) ? ' onload=""' : '' ), $tag );
+			if ( ! empty( $tags_match['media'][ $i ] ) ) {
+				preg_match( '/media\s*=\s*[\'"](?<media>.*)[\'"]/ix', $tags_match['media'][ $i ], $media_match );
+				$original_media = $media_match['media'];
+				$tag            = str_replace( $tags_match['media'][ $i ], 'media="print"', $tag );
+			}
+
+			if ( ! empty( $tags_match['onload'][ $i ] ) ) {
+				$onload_delimiter = substr( $tags_match['onload'][ $i ], 7, 1 );
+				$onload_end       = strpos( $tags_match['onload'][ $i ], $onload_delimiter . ' ' );
+				if ( ! $onload_end ) {
+					$onload_end = strpos( $tags_match['onload'][ $i ], $onload_delimiter . '>' );
+				}
```

Comment:
Can you explain what is happening in this part?

Reply:
The issue with `onload` is that it might contain the quotes: `"` or `'` . So in order to identify the content of onload the only reliable solution which I found is to identify the delimiter:
`onload="` or `onload='` . So the next step will be identify the ending onload delimiter. Which can be either `' ` / `" ` if there are other attributes after onload or onload is the last attribute in that case it is `'>` / `">`.
This will help me to extract the full length onload attribute.

$onload_end will give the length of the onload attribute.

If anyone has another solution I am opened to it.

Regex seemed useless in this case because I was unable to identify the full content of onload if it contained '" inside it.